neos/flow-development-collection
# FEATURE: Lazy injections for typed properties

This change adjusts the proxy building and dependency injection process to properly handle PHP 7.4 class property type declarations.

It resolves #2114
Also resolve #2601 

**How to verify it**

What did *not* work before this change was using property type declarations for lazily injected properties. 

Consider the following code: injecting `$testSingleton` is what this fix makes possible.

```php
<?php
namespace Acme\LazyTest\Command;

use Acme\LazyTest\TestSingleton;
use Neos\Flow\Annotations as Flow;
use Neos\Flow\Cache\CacheManager;
use Neos\Flow\Cli\CommandController;
use Psr\Log\LoggerInterface;

/**
 * @Flow\Scope("singleton")
 */
class TestCommandController extends CommandController
{
    /**
     * LoggerInterface doesn't have to be nullable, because the proxy class
     * builder will not even try to use lazy injection, since it is an interface
     *
     * @Flow\Inject
     **/
    protected LoggerInterface $logger;

    /**
     * CacheManager must be nullable, because it can be injected lazily
     *
     * @Flow\Inject
     */
    protected ?CacheManager $cacheManager;

    /**
     * @Flow\Inject
     */
    protected ?TestSingleton $testSingleton;

    /**
     * Test command
     *
     * @return void
     */
    public function testCommand(): void
    {
        # Will be the actual logger, because an interface is injected,
        # which results in an eager dependency:

        var_dump(get_class($this->logger)); # Neos\Flow\Log\Psr\Logger

        # Will be the actual Cache Manager, because as soon as the
        # Cache Manager is used (instantiated) elsewhere, it is also
        # injected eagerly into this controller:

        var_dump(get_class($this->cacheManager));                      # Neos\Flow\Cache\CacheManager

        # The TestSingleton is not injected into any other class and
        # therefore is a real Lazy Proxy before it is used for the
        # first time by calling hello():

        var_dump(get_class($this->testSingleton));                     # Acme\LazyTest\TestSingleton_LazyProxy
        var_dump($this->testSingleton instanceof TestSingleton); # bool(true)
        $this->testSingleton->hello('world');
        var_dump(get_class($this->testSingleton));                     # Acme\LazyTest\TestSingleton
    }
}
```

In order to verify this change, you can use the test package (see zip file) and play a bit with the injected properties. Things to try out (at least):

- verify the output in Development context after initial run of `./flow test:test` with an empty cache
- verify the output on a second rund
- verify that it works when you make a little change, so that proxy classes are built
- check that all this works in Production

[Acme.LazyTest.zip](https://github.com/neos/flow-development-collection/files/8118718/Acme.LazyTest.zip)
